crates/dbsp/src/circuit/metrics.rs

```diff
@@ -57,9 +61,180 @@ pub const COMPACTION_DURATION: &str = "file.compaction_duration";
+
+impl Histogram {
+    /// Describe and initialize a new [`Histogram`].
+    ///
+    /// The following labels are set to this histogram by default:
+    /// `worker` => Appropriate runtime worker thread index
+    /// `gid` => The global node id of the current operator
+    #[allow(unused)]
+    pub(crate) fn new(
+        name: &str,
+        description: Option<String>,
+        unit: Option<&str>,
+        gid: &GlobalNodeId,
+        mut labels: Vec<(String, String)>,
+    ) -> Self {
+        labels.push(("worker".to_owned(), Runtime::worker_index().to_string()));
+        labels.push(("gid".to_owned(), gid.metrics_id()));
+
+        let unit: Option<metrics::Unit> = unit.and_then(metrics::Unit::from_string);
+        let name = metric_name(name);
+        let description = description.unwrap_or_default();
+
+        match unit {
+            Some(unit) => {
+                describe_histogram!(name.clone(), unit, description);
+            }
+            None => {
+                describe_histogram!(name.clone(), description);
+            }
+        };
+
+        Self(metrics::histogram!(name, &labels[..]))
+    }
+
+    #[allow(unused)]
+    pub(crate) fn record(&self, value: f64) {
+        self.0.record(value)
+    }
+}
+
+impl SizeOf for Histogram {
+    fn size_of_children(&self, _: &mut Context) {}
+}
+
+/// Metric of type `Counter`.
+///
+/// Counters represent a single monotonic value, which means the value can only be incremented,
+/// not decremented, and always starts out with an initial value of zero.
+#[derive(Clone)]
+pub(crate) struct Counter(metrics::Counter);
+
+impl Counter {
+    /// Describe and initialize a new [`Counter`].
+    ///
+    /// The following labels are set to this counter by default:
+    /// `worker` => Appropriate runtime worker thread index
+    /// `gid` => The global node id of the current operator
+    #[allow(unused)]
+    pub(crate) fn new(
+        name: &str,
+        description: Option<String>,
+        unit: Option<&str>,
+        gid: &GlobalNodeId,
+        mut labels: Vec<(String, String)>,
+    ) -> Self {
+        labels.push(("worker".to_owned(), Runtime::worker_index().to_string()));
+        labels.push(("gid".to_owned(), gid.metrics_id()));
+
+        let unit: Option<metrics::Unit> = unit.and_then(metrics::Unit::from_string);
+        let name = metric_name(name);
+        let description = description.unwrap_or_default();
+
+        match unit {
+            Some(unit) => {
+                describe_counter!(name.clone(), unit, description);
+            }
+            None => {
+                describe_counter!(name.clone(), description);
+            }
+        };
+
+        Self(metrics::counter!(name, &labels[..]))
+    }
+
+    /// Increments this [`Counter`] by `value`.
+    #[allow(unused)]
+    pub(crate) fn increment(&self, value: u64) {
+        self.0.increment(value)
+    }
+}
+
+impl SizeOf for Counter {
+    fn size_of_children(&self, _: &mut Context) {}
+}
+
```

Comment:
This was a workaround the size_of trait requirement.

What do you guys think? @gz @ryzhyk

Reply:
Oh, for this it's better to use the `#[size_of(skip)]` annotation on fields that don't implement the `SizeOf` trait. You can find many examples in the code.